fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/search/LuceneOptimizedIndexSearcher.java

```diff
@@ -0,0 +1,115 @@
+import org.apache.lucene.search.CollectionTerminatedException;
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.Weight;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+
+/**
+ * This class optimizes the current IndexSearcher and attempts to perform operations in parallel in places where
+ * data access can occur.
+ *
+ *
+ */
+public class LuceneOptimizedIndexSearcher extends IndexSearcher {
+
+    public LuceneOptimizedIndexSearcher(final IndexReader r) {
+        super(r);
+    }
+
+    public LuceneOptimizedIndexSearcher(final IndexReader r, final Executor executor) {
+        super(r, executor);
+    }
+
+    /**
+     * Lower-level search API.
+     *
+     * This is optimized into an attempt to parallelize database access in FDB
+     *
+     * <p>
+     * {@link LeafCollector#collect(int)} is called for every document. <br>
+     *
+     * <p>
+     * NOTE: this method executes the searches on all given leaves exclusively.
+     * To search across all the searchers leaves use {@link #leafContexts}.
+     *
+     * @param leaves
+     *          the searchers leaves to execute the searches on
+     * @param weight
+     *          to match documents
+     * @param collector
+     *          to receive hits
+     * @throws BooleanQuery.TooManyClauses If a query would exceed
+     *         {@link BooleanQuery#getMaxClauseCount()} clauses.
+     */
+    @Override
+    @SuppressWarnings("PMD")
+    protected void search(List<LeafReaderContext> leaves, Weight weight, Collector collector)
+            throws IOException {
+        Executor executor = getExecutor();
+        if (executor == null) {
+            super.search(leaves, weight, collector);
+        } else {
+            // Used to parallelize weight.bulkScorer database access
+            Map<LeafReaderContext, CompletableFuture<BulkScorer>> ctxScorers =
+                    new ConcurrentHashMap<>();
+            leaves.forEach(ctx -> ctxScorers.put(ctx, CompletableFuture.supplyAsync(() -> {
+                try {
+                    collector.getLeafCollector(ctx);
+                    return weight.bulkScorer(ctx);
+                } catch (IOException e) {
+                    return null;
+                }
+            }, getExecutor())));
+
+            // Need to be able to throw IOExceptions out of this stack.
+            // For example, attempting to perform phrase search without pos/freqs throws out of this
+            // piece of code.
+            for (LeafReaderContext ctx : leaves) {
+                try {
+                    final LeafCollector leafCollector = collector.getLeafCollector(ctx);
+                    BulkScorer scorer = ctxScorers.get(ctx).join();
```

Comment:
I believe join wraps errors from the  in `CompletionException`. It may make sense, if (as the comment suggests) we need to throw IOExceptions to have logic to catch `CompletionException`s and unwrap them